src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs

```diff
@@ -1350,9 +1388,61 @@ public S(int arg)
 ");
             var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
             comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.NotEmpty(comp.GetTypeByMetadataName("S").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
+            comp.VerifyDiagnostics();
+
+            Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
+        }
 
+        [Fact]
+        public void SpeculativeSemanticModelTest()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    public int P { get => 0; }
+}
+");
+            var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
+            comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers(GeneratedNames.MakeBackingFieldName("P")));
             comp.VerifyDiagnostics();
 
+            var tree = comp.SyntaxTrees[0];
+            var token = tree.GetRoot().DescendantTokens().Single(t => t.IsKind(SyntaxKind.NumericLiteralToken));
+
+            var model = comp.GetSemanticModel(comp.SyntaxTrees[0]);
+            var identifier = SyntaxFactory.IdentifierName(SyntaxFactory.Identifier(leading: default, contextualKind: SyntaxKind.FieldKeyword, text: "field", valueText: "field", trailing: default));
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, identifier, out var speculativeModel);
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+
+            var fieldKeywordTypeInfo = speculativeModel.GetTypeInfo(identifier);
+            Assert.Equal(SpecialType.System_Int32, fieldKeywordTypeInfo.Type.SpecialType);
+            Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
+        }
+
+        [Fact]
+        public void SpeculativeSemanticModelTest2()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    public int P { get => 0; }
+}
+");
+            var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
+            comp.TestOnlyCompilationData = accessorBindingData;
+
+            var tree = comp.SyntaxTrees[0];
+            var token = tree.GetRoot().DescendantTokens().Single(t => t.IsKind(SyntaxKind.NumericLiteralToken));
+
+            var model = comp.GetSemanticModel(comp.SyntaxTrees[0]);
+            var identifier = SyntaxFactory.IdentifierName(SyntaxFactory.Identifier(leading: default, contextualKind: SyntaxKind.FieldKeyword, text: "field", valueText: "field", trailing: default));
+            model.TryGetSpeculativeSemanticModel(token.SpanStart, identifier, out var speculativeModel);
+            var fieldKeywordTypeInfo = speculativeModel.GetTypeInfo(identifier);
+
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
```

Comment:
This assertion fails. What I was trying to do in this test is to not cause binding to the original compilation so that the sentinel value remains, then I only bind the speculated node. This caused the property to have a backing field, which `GetFieldsToEmit` will return. @AlekseyTs Any pointers as to what I need to do, for both this test and the previous one?

Reply:
> Any pointers as to what I need to do, for both this test and the previous one?

I agree that the behavior doesn't feel right. But before doing anything to fix it, let's add more tests to better understand the set of scenarios we should handle.

Note, the tests should pass, asserts should reflect the current behavior. Otherwise, reviewers cannot "see" the current behavior. If some behavior feel unexpected to you, add a comment.

Here are additional tests that I suggest:
- Speculate with a field identifier within a an accessor that uses the field keyword. 
- Speculate with a field identifier within a an accessor that doesn't use the field keyword, but there is another accessor that uses the keyword.
- Speculate in a duplicate accessor (in the second getter, for example). Both when the first accessor uses and doesn't use the field keyword.

Once we have these tests, I might suggest more scenarios to try.

Reply:
@AlekseyTs For each of those scenarios, do you want two tests:

1. Bind original tree first
2. Bind speculated tree first

as I demonstrated in https://github.com/dotnet/roslyn/pull/59270#discussion_r800035455

If not, which scenario of the two is the interesting one?

Reply:
> For each of those scenarios, do you want two tests?

Yes, let's have two a pair of tests per scenario.